### ISSUE TITLE
Installation does not work with python3-based pip

### ISSUE THREAD

**Issue by maweki** [johnyf/dd#3]
As it says in the title.

    $ pip --version
    pip 7.1.2 from /home/maweki/.local/lib/python3.4/site-packages (python 3.4
    $ python --version
    Python 2.7.10

when I try to install dd via `pip install --user dd` I get the following error 

  Collecting dd
  Using cached dd-0.1.3.tar.gz
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 20, in <module>
      File "/tmp/pip-build-ew5h83vp/dd/setup.py", line 4, in <module>
        import download
      File "/tmp/pip-build-ew5h83vp/dd/download.py", line 7, in <module>
        import urllib2
    ImportError: No module named 'urllib2'
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-ew5h83vp/dd

It does work with `pip2 install --user dd` though. So maybe this package should not be available for python3 maybe

**Comment by slivingston** [comment]
[pydot as on PyPI](https://pypi.python.org/pypi/pydot/1.0.28) does not work with Python 3. I found [a fork that claims Python 3 compatibility](https://github.com/nlhepler/pydot).

**Comment by maweki** [comment]
Pydot isn't a strict requirement for installing. The error was the downloading that was fixed in my branch/pull-request (https://github.com/johnyf/dd/pull/4)

**Comment by slivingston** [comment]
It is not strict in that an "extra" must be requested, e.g., as in `pip install dd[dot]`. My comment about pydot compatibility is relevant for this issue insofar as we want `dd` to be available via `pip` in both Python 2 and 3. The OP does not preclude the extras.

**Comment by maweki** [comment]
dd is available via pypi for python3 because no supported versions are set within pypi.
dd in general doesn't work with python3 which is (probably) a rather easy fix. 

But we won't do it because one optional dependency does not support python3? Again ignoring the fact that one can allready (try to) install dd for python3.

I'd think, we'd want to do these things in no particular order:

* make dd (core) python3 compatible
* set the correct supported versions in pypi
* fix pydot, adding python3 support

**Comment by slivingston** [comment]
I agree that core should be made compatible with Python 3 and then, if @johnyf prefers, released but without supporting the "dot" extra on Python 3. I.e., I am not arguing that @johnyf  should wait on dealing with pydot before moving forward with Python 3. [My comment](https://github.com/johnyf/dd/issues/3#issuecomment-151985083) was intended to indicate that this issue should stay open until that optional dependency is addressed.

I support abandoning pydot entirely and can contribute a replacement for desired functions.

**Comment by johnyf** [comment]
I agree that `dd` can be updated relatively soon to work in Python 3 (the Cython bindings are expected to already do so, due to the Cython transpiler addressing [both Python versions from the same source](https://github.com/cython/cython/wiki/FAQ#id13)).
Branch `dev` currently contains this candidate update.

About `pydot`, it cannot be installed from PyPI (at least not w/o a lot of security-overriding options to `pip`, if that's still possible). So, the only installation is possible separately. Therefore, the user can select to install a Python 3 compatible fork, with the same name. There are several [`pydot` forks on already PyPI](https://pypi.python.org/pypi?%3Aaction=search&term=pydot&submit=search).

I will likely contact the author of `pydot` and ask about the possibility of adding a maintainer, on both GitHub and PyPI. But that is an independent issue. If that fails, then I think it is time to use a fork.

**Comment by johnyf** [comment]
The author of `pydot` appears to have a PyPI entry called [`pydot3k`](https://pypi.python.org/pypi/pydot3k/1.0.17) that installs a `pydot` version that works in Python 3. So, for now, it should suffice to conditionally add `pydot` or `pydot3k` to the `setup.py` requirements, depending on the detected Python version.

**Comment by johnyf** [comment]
Unlike the `pydot` entry, the `pydot3k` entry *is* uploaded on PyPI.
So, the claim [above](https://github.com/johnyf/dd/issues/3#issuecomment-152005412) that pydot “cannot be installed from PyPI” does not hold -- at least not for Python 3.

**Comment by johnyf** [comment]
Cython appears to support some aspects of Python 2 and 3, but not others. In particular, string conversion [from Python string objects to Cython](http://docs.cython.org/src/tutorial/strings.html#encoding-text-to-bytes) that works in both Python 2 and 3 is tricky and fragile (and was a pain to achieve).

**Comment by johnyf** [comment]
Conditional dependency on `pydot` or `pydot3k` introduced in a911167581ba0b5d231a700e16859ca2f25a970c, and fewer unnecessary calls to `pydot` in 3bb95afdc98faf73eaf9d03a7ce5c6b67837c9af.

**Comment by johnyf** [comment]
#4, #5 and #6 address Python 3 compatibility of `dd` (item 1 [above](https://github.com/johnyf/dd/issues/3#issuecomment-151992596)).

**Comment by johnyf** [comment]
1599c51ea18f3da218e8d5ed15d091ddf2091425 adds language classifiers on PyPI (item 2 [above](https://github.com/johnyf/dd/issues/3#issuecomment-151992596)).
